packages/datadog-instrumentations/src/rhea.js

```diff
@@ -0,0 +1,215 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const circularBufferConstructor = Symbol('circularBufferConstructor')
+const dd = Symbol('datadog')
+const inFlightDeliveries = Symbol('inFlightDeliveries')
+
+const patched = new WeakSet()
+const dispatchCh = channel('apm:rhea:general:dispatch')
+const errorCh = channel('apm:rhea:general:error')
+const asyncEndCh = channel('apm:rhea:general:async-end')
+const endCh = channel('apm:rhea:general:end')
+
+addHook({ name: 'rhea', versions: ['>=1'], file: 'lib/link.js' }, obj => {
+  const startSendCh = channel('apm:rhea:send:start')
+  const startReceiveCh = channel('apm:rhea:receive:start')
+
+  const Sender = obj.Sender
+  const Receiver = obj.Receiver
+  shimmer.wrap(Sender.prototype, 'send', send => function (msg, tag, format) {
+    if (!canTrace(this)) {
+      // we can't handle disconnects or ending spans, so we can't safely instrument
+      return send.apply(this, arguments)
+    }
+
+    const name = getResourceNameFromSender(this)
+    const { host, port } = getHostAndPort(this.connection)
+
+    startSendCh.publish({ name, host, port, msg })
+    const delivery = send.apply(this, arguments)
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+    delivery[dd] = { asyncResource }
+    addToInFlightDeliveries(this.connection, delivery)
+    try {
+      return delivery
+    } catch (err) {
+      errorCh.publish(err)
+
+      throw err
+    } finally {
+      endCh.publish(undefined)
+    }
+  })
+
+  shimmer.wrap(Receiver.prototype, 'dispatch', dispatch => function (eventName, msgObj) {
+    if (!canTrace(this)) {
+      // we can't handle disconnects or ending spans, so we can't safely instrument
+      return dispatch.apply(this, arguments)
+    }
+
+    if (eventName === 'message' && msgObj) {
+      startReceiveCh.publish({ msgObj, connection: this.connection })
+
+      if (msgObj.delivery) {
+        const asyncResource = new AsyncResource('bound-anonymous-fn')
+        msgObj.delivery[dd] = { asyncResource }
```

Comment:
I removed this kind of reference from the code base a couple of weeks ago but it seems I forgot this one. We should never add any properties to library objects. A WeakMap can be used instead.